### PR TITLE
feat(material-experimental): add test harness for slider

### PR DESCRIPTION
**Pull request by devversion**

### CODE REVIEW

**Comment by devversion** [comment]
@jelbourn @mmalerba I wasn't too sure how we can solve the `setValue` thing. There are two options:

1. Setting the value through click (this is what currently is implemented as part of https://github.com/angular/components/pull/16688/commits/51e4daa2527ea1f47569d54ec9631f5e0da1815f)
     * Can deviate in rare cases since not each value maps to a single pixel (that we an click).
2. Setting the value by pressing the up or down arrow keys. This is unpredictable since the keys always cause a change based on the specified `mat-slider#step`.
     *  Developer could specify a value that can't be selected through keyboard but got set through binding for example.

**Comment by mmalerba** [comment]
Ah that's interesting... I guess a real user would have similar difficulty using your slider if you didn't have it set so that all values can be reached via the keyboard or had more steps than pixels worth of track.

I guess I would say go for the clicking approach like you've done and maybe document that if you have more steps than pixels the `setValue` could be slightly off, but also that it would be a bad experience for your users and you should change it

**Comment by devversion** [comment]
@mmalerba Done. I'm wondering if we should consider providing a method for selecting a value through keyboard interaction too. The arrow keys allow granular control over the value since those allow changes by _"one step value"_ but in order to provide meaningful methods.. we'd need to somehow determine the step value from within the harness (which requires some host bindings on the slider..). I think that's something we could explore in a follow-up PR.

File: src/material-experimental/mdc-slider/harness/slider-harness.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  /** Gets the orientation of the slider. */
+  async getOrientation(): Promise<'horizontal'|'vertical'> {
+    // "aria-orientation" will always be set to either "horizontal" or "vertical".
+    return (await this.host()).getAttribute('aria-orientation') as any;
+  }
+
+  /**
+   * Sets the value of the slide by clicking on the slider track.
```
Comment (jelbourn) [inline src/material-experimental/mdc-slider/harness/slider-harness.ts]: "slide" -> "slider"

File: src/material/slider/slider.ts [hunk 2]
```diff
@@ -132,6 +132,7 @@ const _MatSliderMixinBase:
     '[class.mat-slider-has-ticks]': 'tickInterval',
     '[class.mat-slider-horizontal]': '!vertical',
     '[class.mat-slider-axis-inverted]': '_invertAxis',
+    '[class.mat-slider-invert-mouse-coords]': '_shouldInvertMouseCoords()',
```
Comment (jelbourn) [inline src/material/slider/slider.ts]: Can you add a comment that this class is only used for the harness?

**Comment by jelbourn** [comment]
I also agree that methods to increment/decrement by the step via keyboard would be good in a follow-up.

**Comment by devversion** [comment]
@jelbourn @mmalerba Addressed feedback.

File: src/material-experimental/mdc-slider/harness/slider-harness.spec.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    expect(await sliders[2].getValue()).toBe(225);
+
+    await sliders[1].setValue(33);
+    await sliders[2].setValue(300);
+
+    expect(await sliders[1].getValue()).toBe(33);
+    // value should be clamped to the maximum.
+    expect(await sliders[2].getValue()).toBe(250);
+
+    // should not retrieve incorrect values in case slider is inverted
```
Comment (mmalerba) [inline src/material-experimental/mdc-slider/harness/slider-harness.spec.ts]: can probably make this a separate test

File: src/material-experimental/mdc-slider/harness/slider-harness.spec.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    // due to RTL page layout.
+    fixture.componentInstance.dir = 'rtl';
+    fixture.detectChanges();
+
+    await sliders[1].setValue(80);
+    expect(await sliders[1].getValue()).toBe(80);
+
+    // should not retrieve incorrect values in case sliders are inverted.
```
Comment (mmalerba) [inline src/material-experimental/mdc-slider/harness/slider-harness.spec.ts]: this too

**Pull request merged**